azure-iot-device/tests/common/pipeline/pipeline_stage_test.py

```diff
@@ -14,29 +14,202 @@
+        @pytest.mark.it("Completes the op with failure (PipelineError) if there is no next stage")
+        def test_fails_op_when_no_next_stage(self, mocker, stage, arbitrary_op):
+            stage.next = None
+
+            assert not arbitrary_op.completed
+
+            stage.send_op_down(arbitrary_op)
+
+            assert arbitrary_op.completed
+            assert type(arbitrary_op.error) is pipeline_exceptions.PipelineError
+
+        @pytest.mark.it("Passes the op to the next stage's .run_op() method")
+        def test_passes_op_to_next_stage(self, mocker, stage, arbitrary_op):
+            stage.send_op_down(arbitrary_op)
+            assert stage.next.run_op.call_count == 1
+            assert stage.next.run_op.call_args == mocker.call(arbitrary_op)
+
+    @pytest.mark.describe("{} - .send_event_up()".format(stage_class_under_test.__name__))
+    class StageSendEventUpTests(StageTestConfig):
+        @pytest.mark.it(
+            "Passes the event up to the previous stage's .handle_pipeline_event() method"
+        )
+        def test_calls_handle_pipeline_event(self, stage, arbitrary_event, mocker):
+            stage.send_event_up(arbitrary_event)
+            assert stage.previous.handle_pipeline_event.call_count == 1
+            assert stage.previous.handle_pipeline_event.call_args == mocker.call(arbitrary_event)
+
+        @pytest.mark.it(
+            "Sends a PipelineError to the background exception handler instead of sending the event up the pipeline, if there is no previous pipeline stage"
+        )
+        def test_no_previous_stage(self, stage, arbitrary_event, mocker):
+            stage.previous = None
+            mocker.spy(handle_exceptions, "handle_background_exception")
+
+            stage.send_event_up(arbitrary_event)
+
+            assert handle_exceptions.handle_background_exception.call_count == 1
+            assert (
+                type(handle_exceptions.handle_background_exception.call_args[0][0])
+                == pipeline_exceptions.PipelineError
+            )
+
+    setattr(
+        test_module,
+        "Test{}SendOpDown".format(stage_class_under_test.__name__),
+        StageSendOpDownTests,
+    )
+    setattr(
+        test_module,
+        "Test{}SendEventUp".format(stage_class_under_test.__name__),
+        StageSendEventUpTests,
+    )
+
+    #############################################
+    # RUN OP / HANDLE_PIPELINE_EVENT BASE TESTS #
+    #############################################
+
+    # These tests are only run if the Stage in question has NOT overridden the PipelineStage base
+    # implementations of ._run_op() and/or ._handle_pipeline_event()
+
+    if stage_class_under_test._run_op is PipelineStage._run_op:
+
+        @pytest.mark.describe("{} - .run_op()".format(stage_class_under_test.__name__))
+        class StageRunOpUnhandledOp(StageTestConfig, StageRunOpTestBase):
+            @pytest.fixture
+            def op(self, arbitrary_op):
```

Comment:
>self [](start = 19, length = 4)

this fixture seems unnecessary, if it's just returning an arbitrary_op

Reply:
It alters the name

---
In reply to: [355036486](https://github.com/Azure/azure-iot-sdk-python/pull/409#discussion_r355036486) [](ancestors = 355036486)